rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/ExtendedFuture.java

```diff
@@ -0,0 +1,95 @@
+package org.rascalmpl.vscode.lsp.util;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A wrapper around CompletableFuture's that allow for us to replace the results
+ * of a still running future, and call an specific closure if it is replaced by
+ * a newer version.
+ */
+public class ExtendedFuture<T> {
+    private static final Logger logger = LogManager.getLogger(ExtendedFuture.class);
+
+    private static final class InterruptibleFuture<T> {
+        private final Runnable interrupt;
+        private final CompletableFuture<T> future;
+
+        private InterruptibleFuture(Runnable interrupt, CompletableFuture<T> future) {
+            this.interrupt = interrupt;
+            this.future = future;
+        }
+
+    }
+
+    private final AtomicBoolean valid;
+    private final AtomicReference<InterruptibleFuture<T>> actual;
+
+    public ExtendedFuture(CompletableFuture<T> start, Runnable interrupt) {
+        actual = new AtomicReference<>(new InterruptibleFuture<>(interrupt, start));
+        valid = new AtomicBoolean(true);
+    }
+
+    public CompletableFuture<T> get() {
+        return actual.get().future;
+    }
+
+    public boolean isValid() {
+        return valid.get();
+    }
+
+    public void invalidate() {
+        valid.set(false);
+    }
+
+    @SuppressWarnings("java:S1181")
+    public <U> CompletableFuture<T> replace(CompletableFuture<U> base, Function<U, T> apply, Runnable interrupt,
```

Comment:
I can use documentation for the parameters; I don't understand their function from their names. Also a description of the functionality of replace in general is needed in javadoc.